cmd/promtool/backfill_test.go

```diff
@@ -0,0 +1,343 @@
+			allSamples = append(allSamples, series[0])
+			return nil
+		}()
+		require.NoError(t, err)
+	}
+	allSymbolsSlice := make([]string, 0)
+	for key := range allSymbols {
+		allSymbolsSlice = append(allSymbolsSlice, key)
+	}
+	sort.Strings(allSymbolsSlice)
+	sort.Strings(expectedSymbols)
+	require.Equal(t, expectedSymbols, allSymbolsSlice)
+	sortSamples(allSamples)
+	sortSamples(expectedSamples)
+	require.Equal(t, expectedSamples, allSamples)
+	require.Equal(t, expectedMinTime, mint)
+	require.Equal(t, expectedMaxTime, maxt)
+}
+
+func TestBackfill(t *testing.T) {
+	tests := []struct {
+		ToParse      string
+		IsOk         bool
+		MetricLabels []string
+		Description  string
+		Expected     struct {
+			MinTime   int64
+			MaxTime   int64
+			Symbols   []string
+			NumBlocks int
+			Samples   []backfillSample
+		}
+	}{
+		{
+			ToParse:     `# EOF`,
+			IsOk:        true,
+			Description: "Empty file.",
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1565133713989
+http_requests_total{code="400"} 1 1565133713990
+# EOF
+`,
+			IsOk:        true,
+			Description: "Multiple samples with different timestamp for different series.",
+			Expected: struct {
+				MinTime   int64
+				MaxTime   int64
+				Symbols   []string
+				NumBlocks int
+				Samples   []backfillSample
+			}{
+				MinTime:   1565133713989000,
+				MaxTime:   1565133713990001,
+				Symbols:   []string{"http_requests_total", "code", "200", "400", "__name__"},
+				NumBlocks: 1,
+				Samples: []backfillSample{
+					{
+						Timestamp: 1565133713989000,
+						Value:     1021,
+						Labels:    labels.FromStrings("__name__", "http_requests_total", "code", "200"),
+					},
+				},
+			},
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1565133713989
+http_requests_total{code="200"} 1 1565133714989
+http_requests_total{code="400"} 2 1565133715989
+# EOF
+`,
+			IsOk:         true,
+			Description:  "Multiple samples with different timestamp for the same series",
+			MetricLabels: []string{"__name__", "http_requests_total"},
+			Expected: struct {
+				MinTime   int64
+				MaxTime   int64
+				Symbols   []string
+				NumBlocks int
+				Samples   []backfillSample
+			}{
+				MinTime:   1565133713989000,
+				MaxTime:   1565133715989001,
+				Symbols:   []string{"http_requests_total", "code", "200", "400", "__name__"},
+				NumBlocks: 1,
+				Samples: []backfillSample{
+					{
+						Timestamp: 1565133713989000,
+						Value:     1021,
+						Labels:    labels.FromStrings("__name__", "http_requests_total", "code", "200"),
+					},
+				},
+			},
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1565133713989
+http_requests_total{code="400"} 1 1565144513989
```

Comment:
Those two samples end up in different blocks, but they are also for different series.

Could you add another sample for each series so that we have two series, each spanning two blocks?

Reply:
yes sure